### PR TITLE
feat(catalog-backend): reject almost-same-name entities

### PR DESCRIPTION
**Pull request by freben**
We promised this functionality, but it hasn't actually been enforced until now.

### CODE REVIEW

File: plugins/catalog-backend/src/database/CommonDatabase.test.ts [hunk 1]
```diff
@@ -14,39 +14,23 @@
...
-      connection: ':memory:',
-      useNullAsDefault: true,
-    });
-
-    await knex.raw('PRAGMA foreign_keys = ON');
-    await knex.migrate.latest({
-      directory: path.resolve(__dirname, 'migrations'),
-      loadExtensions: ['.ts'],
-    });
+    db = await DatabaseManager.createTestDatabase();
```
Comment (shmidt-i) [inline plugins/catalog-backend/src/database/CommonDatabase.test.ts:33]: nice 😎

**Review by Nek — APPROVED** [review]
LGTM  🥇

File: plugins/catalog-backend/src/database/types.ts [hunk 2]
```diff
@@ -23,6 +23,9 @@ export type DbEntitiesRow = {
   kind: string;
   name: string | null;
   namespace: string | null;
+  kind_normalized: string;
```
Comment (shmidt-i) [inline plugins/catalog-backend/src/database/types.ts:26]: Should we add them here as `| null` since they are nullable in the table?

**Review by shmidt-i — APPROVED** [review]

**Comment by Rugvip** [comment]
Could we not just do the validation by fetching all entities from the DB for now? Feels like it's unfortunate to bring this into the DB schema already

**Comment by freben** [comment]
@Rugvip only if we hold a write exclusive lock, right

**Comment by Rugvip** [comment]
I don't see why we'd care about write consistency at this point

File: plugins/catalog-backend/src/database/types.ts (hunk 2 above)
Comment (freben) [reply plugins/catalog-backend/src/database/types.ts:26]: Was not sure if best to reflect db reality or model reality :) but sure I could change that

**Comment by freben** [comment]
I don't see why not? Being lax and permitting bad states already now sounds like a poor start to things :)

**Comment by freben** [comment]
And I would have to do it in several places (add, update loops, future patch ability) as well instead of having the db strongly enforce it at all times.

**Comment by Rugvip** [comment]
Because it makes the implementation a lot more difficult to change, and we're at a point where we want to be able to do experiments and changes easily. And you'd still just do it in the DB layer?

**Comment by freben** [comment]
Do you feel that schema migrations are hard? Because this is less code and stronger guarantees and just the kind of thing that databases excel at (at little-to-no implementation cost to us).

**Comment by Rugvip** [comment]
Yes, extending the database schema adds more complexity than adding a function that's called in a couple of places

**Review by Rugvip — APPROVED** [review]
🎉  🙏

**Pull request merged**